### PR TITLE
[Merged by Bors] - feat(Data/FinEnum): Option instance and Type recursor

### PR DESCRIPTION
**Pull request by TpmKranz**
Replicate `Data/Fintype/Option`, but yield non-truncated data.

First chunk of #12648

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/FinEnum/Option.lean [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+/-- A recursor principle for finite-and-enumerable types, analogous to `Nat.recOn`.
+It effectively says that every `FinEnum` is either `Empty` or `Option α`, up to an `Equiv`.
+In contrast to the `Fintype` case, data can be transported along such an `Equiv`. -/
+def recOnEmptyOption {P : (α : Type u) → [FinEnum α] → Sort v}
+    {α : Type u} (aenum : FinEnum α)
+    (of_equiv : {α β : Type u} → [FinEnum α] → [FinEnum β] → α ≃ β → P α → P β)
+    (h_empty : P PEmpty.{u + 1})
+    (h_option : {α : Type u} → [FinEnum α] → P α → P (Option α)) :
+    P α := @recEmptyOption P of_equiv h_empty h_option α aenum
+end FinEnum
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:68]:
```suggestion

end FinEnum
```

File: Mathlib/Data/FinEnum/Option.lean [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        Equiv.trans (@finSuccEquivLast n) <| Equiv.optionCongr Equiv.ulift.symm
+    have opsumeq := Equiv.optionEquivSumPUnit.{u,u} <| ULift (Fin n)
+    have fssumeq := Equiv.trans optioeq opsumeq
+    have fssumfe := @FinEnum.sum _ _ fe FinEnum.punit.{u}
+    have fsuccfe := FinEnum.ofEquiv _ fssumeq
+    exact of_equiv fsucceq <| of_equiv optioeq.symm <| h_option <| @ih _ fe fecardeq
+
+/-- A recursor principle for finite-and-enumerable types, analogous to `Nat.recOn`.
+It effectively says that every `FinEnum` is either `Empty` or `Option α`, up to an `Equiv`.
+In contrast to the `Fintype` case, data can be transported along such an `Equiv`. -/
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:61]: Can this cross-ref `Fintype.induction_empty_option` and `Fintype.truncRecEmptyOption` (or maybe the whole `Data.Fintype.Option` file)?

**Comment by kbuzzard** [comment]
This PR is defining data using tactics. To prove that the resulting declarations are usable, can you prove the basic equation lemmas for them, i.e. write that API showing that the definitions produce what you expect them to produce for the empty and option inputs?

**Comment by TpmKranz** [comment]
> This PR is defining data using tactics. To prove that the resulting declarations are usable, can you prove the basic equation lemmas for them, i.e. write that API showing that the definitions produce what you expect them to produce for the empty and option inputs?

Thank you for bringing this to my attention! I've re-written the definition, but since it's a bit awkward to reduce, the lemmas you mentioned are now provided, as well. I would've loved to state them in terms of `IsEmpty` and `NonEmpty`, but that would've made the required proof arguments on the RHSs even more bloated. Since I didn't want those derivations to go to waste, though, I've added them to `FinEnum.lean`. I fear that I might've gone a bit overboard in providing `FinEnum` instances for all `IsEmpty` and `Unique` types, as that's what leads to `simp` now finding more lemmas via typeclass inference.

Are there any recommendations on how to deal with those? Or should I just scrap the generic instances? I don't think the remark that accompanies [https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Fintype/Basic.html#Fintype.ofIsEmpty](`Fintype.ofIsEmpty`) is applicable here, due to `IsEmpty α → Unique (FinEnum α)` implying there cannot be two observably conflicting `FinEnum α` instances. Though I would suppose the same goes for `Fintype α`, so maybe I don't understand the remark thoroughly enough…

**Comment by TpmKranz** [comment]
> You have correctly argued that there couldn't be two propositionally distinct instances of `FinEnum α` when `IsEmpty α` (or more generally `Subsingleton α`), but there could still _definitionally_ distinct instances. Try checking whether the two instances for `Fin 0` (the one coming from `IsEmpty (Fin 0)` and the one coming from `Fin n`) are definitionally equal by doing `example : the_first_instance = the_second_instance := rfl`.

Naturally, the two instances were not definitionally equal. Although I believe that the instances defined here are particularly inelegant, not even

```lean
#check (rfl : ofIsEmpty (α := Fin 0) = ⟨0, Equiv.refl (Fin 0)⟩)
```

goes through. Anyway, the instances are now not haphazardously registered anymore.

File: Mathlib/Data/FinEnum/Option.lean [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+by `Fin`s of equal cardinality.
+In contrast to the `Fintype` case, data can be transported along such an `Equiv`.
+Also, since order matters, the choice of element that gets replaced by `Option.none` has
+to be provided for every step.
+
+Since every `FinEnum` instance implies a `Fintype` instance and `Prop` is squashed already,
+`Fintype.induction_empty_option` can be used if a `Prop` needs to be constructed.
+Cf. `Data.Fintype.Option`
+-/
+def recEmptyOption {P : (α : Type u) → Sort v}
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:43]:
```suggestion
def recEmptyOption {P : Type u → Sort v}
```

File: Mathlib/Data/FinEnum/Option.lean [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    (fin_choice : (n : ℕ) → Fin (n + 1))
+    (of_equiv : {α β : Type u} → (_ : FinEnum α) → (_ : FinEnum β) → card β = card α → P α → P β )
+    (h_empty : P PEmpty.{u + 1})
+    (h_option : {α : Type u} → FinEnum α → P α → P (Option α))
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:47]:
```suggestion
    (finChoice : (n : ℕ) → Fin (n + 1))
    (congr : {α β : Type u} → (_ : FinEnum α) → (_ : FinEnum β) → card β = card α → P α → P β )
    (empty : P PEmpty.{u + 1})
    (option : {α : Type u} → FinEnum α → P α → P (Option α))
```
to match the naming convention

File: Mathlib/Data/FinEnum/Option.lean [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  | n + 1 =>
+    let fN := uliftId (α := Fin n)
+    have : card (ULift.{u} <| Fin n) = n := card_ulift.trans card_fin
+    of_equiv (insertNone _ <| fin_choice n) _
+      (cardeq.trans <| congrArg Nat.succ this.symm) <|
+        h_option fN (recEmptyOption fin_choice of_equiv h_empty h_option _)
+termination_by card α
+
+/--
+For a type whose `card` disappears, the recursion principle evaluates to whatever `of_equiv`
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:61]:
```suggestion
For an empty type, the recursion principle evaluates to whatever `of_equiv`
```

File: Mathlib/Data/FinEnum.lean [hunk 6]
```diff
@@ -56,6 +56,54 @@ def ofList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) : FinEnum
 def toList (α) [FinEnum α] : List α :=
   (List.finRange (card α)).map (equiv).symm
 
+/-- Any two enumerations of the same type have the same length. -/
+theorem card_eq_card {α : Type u} (e₁ e₂ : FinEnum α) : e₁.card = e₂.card :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:60]:
```suggestion
theorem card_unique {α : Type u} (e₁ e₂ : FinEnum α) : e₁.card = e₂.card :=
```
maybe?

File: Mathlib/Data/FinEnum.lean [hunk 7]
```diff
@@ -56,6 +56,54 @@ def ofList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) : FinEnum
...
+  Fin.equiv_iff_eq.mp ⟨e₁.equiv.symm.trans e₂.equiv⟩
+
+/-- Any enumeration of an empty type has length 0. -/
+theorem card_eq_zero_of_IsEmpty {α : Type u} [FinEnum α] [IsEmpty α] : card α = 0 :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:64]:
```suggestion
theorem card_eq_zero {α : Type u} [FinEnum α] [IsEmpty α] : card α = 0 :=
```
to match `Fintype.card_eq_zero`

Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:64]: Can you also write the analogue of `Fintype.card_eq_zero_iff`?

File: Mathlib/Data/FinEnum.lean [hunk 8]
```diff
@@ -56,6 +56,54 @@ def ofList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) : FinEnum
...
+  match h : card α with
+  | 0 => rfl
+  | _ + 1 => ‹IsEmpty α›.elim ((h ▸ equiv).symm 0)
+
+/-- Any enumeration of a non-empty type has a length with an actual predecessor. -/
+def cardPredOfNonempty (α : Type u) [FinEnum α] [Nonempty α] : { n : ℕ // card α = n + 1 } :=
+  match h : card α with
+  | 0 => False.elim <| ‹Nonempty α›.elim ((h ▸ equiv) · |>.elim0)
+  | n + 1 => ⟨n, rfl⟩
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:73]: This is a bit weird given that the predecessor is explicitly given as `card α - 1`. The idiomatic way of finding the predecessor of `a` is not to inhabit `{n : ℕ // a = n + 1}` but rather to prove `0 < a` and use `a - 1`. Hence I would suggest
```suggestion
lemma card_pos {α : Type*} [FinEnum α] [Nonempty α] : 0 < card α := sorry
lemma card_ne_zero {α : Type*} [FinEnum α] [Nonempty α] : card α ≠ 0 := card_pos.ne'
```

File: Mathlib/Data/FinEnum.lean [hunk 9]
```diff
@@ -56,6 +56,54 @@ def ofList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) : FinEnum
...
+
+/-- Any enumeration of a type with unique inhabitant has length 1. -/
+theorem card_eq_one_of_Unique (α : Type u) [FinEnum α] [Unique α] : card α = 1 :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:76]:
```suggestion
theorem card_eq_one (α : Type u) [FinEnum α] [Unique α] : card α = 1 :=
```
Annoyingly, the analogous lemma here is the misnamed `Fintype.card_unique`...

**Review by YaelDillies — COMMENTED** [review]
Generally, can you try deducing your lemmas from the corresponding `Fintype` ones? We have `FinEnum.instFintype` (which should really be called `FinEnum.toFintype`)

File: Mathlib/Data/FinEnum.lean [hunk 10]
```diff
@@ -80,6 +81,15 @@ noncomputable def ofInjective {α β} (f : α → β) [DecidableEq α] [FinEnum
       use f x
       simp only [h, Function.partialInv_left])
 
+instance uliftId [FinEnum α] : FinEnum (ULift α) :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:84]:
```suggestion
instance _root_.ULift.instFinEnum [FinEnum α] : FinEnum (ULift α) :=
```

File: Mathlib/Data/FinEnum.lean [hunk 11]
```diff
@@ -80,6 +81,15 @@ noncomputable def ofInjective {α β} (f : α → β) [DecidableEq α] [FinEnum
...
+  ⟨card α, Equiv.ulift.trans equiv⟩
+
+theorem card_ulift [FinEnum (ULift α)] [FinEnum α] : card (ULift α) = card α :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:87]:
```suggestion
@[simp]
theorem card_ulift [FinEnum (ULift α)] [FinEnum α] : card (ULift α) = card α :=
```

File: Mathlib/Data/FinEnum.lean [hunk 12]
```diff
@@ -80,6 +81,15 @@ noncomputable def ofInjective {α β} (f : α → β) [DecidableEq α] [FinEnum
...
+  Fin.equiv_iff_eq.mp ⟨equiv.symm.trans Equiv.ulift |>.trans equiv⟩
+
+theorem uliftId_equiv [FinEnum α] (i : Fin (card α)) :
+    equiv.symm i = (uliftId.equiv.symm i).down := rfl
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:91]:
```suggestion
section ULift
variable [FinEnum α] (i : Fin (card α))

@[simp] lemma up_equiv : ULift.up (equiv i) = equiv i := rfl
@[simp] lemma down_equiv : (equiv i).down = equiv i := rfl
@[simp] lemma up_equiv_symm : ULift.up (equiv.symm i) = equiv.symm i := rfl
@[simp] lemma down_equiv_symm : (equiv.symm i).down = equiv.symm i := rfl

end ULift
```

File: Mathlib/Data/FinEnum.lean [hunk 13]
```diff
@@ -98,6 +108,8 @@ instance sum {β} [FinEnum α] [FinEnum β] : FinEnum (α ⊕ β) :=
 instance fin {n} : FinEnum (Fin n) :=
   ofList (List.finRange _) (by simp)
 
+theorem card_fin {n} [FinEnum (Fin n)] : card (Fin n) = n := Fin.equiv_iff_eq.mp ⟨equiv.symm⟩
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:111]:
```suggestion
@[simp]
theorem card_fin {n} [FinEnum (Fin n)] : card (Fin n) = n := Fin.equiv_iff_eq.mp ⟨equiv.symm⟩
```

File: Mathlib/Data/FinEnum.lean [hunk 14]
```diff
@@ -156,6 +168,71 @@ instance (priority := 100) [FinEnum α] : Fintype α where
   elems := univ.map (equiv).symm.toEmbedding
   complete := by intros; simp
 
+/-- The enumeration merely adds an ordering, leaving the cardinality as is. -/
+theorem card_eq_fintype_card {α : Type u} [FinEnum α] [Fintype α] : card α = Fintype.card α :=
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum.lean:172]:
```suggestion
theorem card_eq_fintypeCard {α : Type u} [FinEnum α] [Fintype α] : card α = Fintype.card α :=
```

File: Mathlib/Data/FinEnum/Option.lean [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+/-- Inserting an `Option.none` anywhere in an enumeration yields another enumeration. -/
+def insertNone (α : Type u) [FinEnum α] (i : Fin (card α + 1)) : FinEnum (Option α) where
+  card := card α + 1
+  equiv := equiv.optionCongr.trans <| finSuccEquiv' i |>.symm
+
+/-- This is an arbitrary choice of insertion rank for a default instance.
+It keeps the mapping of the existing `α`-inhabitants intact, modulo `Fin.castSucc`. -/
+instance instFinEnumOptionLast (α : Type u) [FinEnum α] : FinEnum (Option α) :=
+  insertNone α (Fin.last <| card α)
```
Comment (YaelDillies) [inline Mathlib/Data/FinEnum/Option.lean:30]:
```suggestion
  insertNone α (Fin.last _)
```

**Review by YaelDillies — APPROVED** [review]
Thank you for the PR!

maintainer delegate

**Comment by YaelDillies** [comment]
I would avoid mentioning `ULift.instFinEnum` directly. Apart from that, 👍

**Comment by TpmKranz** [comment]
Done. I've also changed the names of the first two ULift simp lemmata to what I believe reflects the naming conventions after the interpreting I did.

**Review by b-mehta — COMMENTED** [review]
Thanks!

bors merge

**Pull request closed without merging**